sdk/core/core-rest-pipeline/package.json

```
@@ -3,8 +3,16 @@
   "version": "1.9.1",
   "description": "Isomorphic client library for making HTTP requests in node.js and browser.",
   "sdk-type": "client",
-  "main": "dist/index.js",
+  "main": "dist/index.cjs",
   "module": "dist-esm/src/index.js",
+  "type": "module",
+  "export": {
+    ".": {
+      "types": "./types/src/index.js",
```

wouldn't this still be something like
```suggestion
      "types": "./types/latest/src/index.d.ts",
```

Though really shouldn't this match the main types entry so `core-rest-pipeline.shims.d.ts` ?

---

Copied from Matt's PR. Will fix.

---

@mpodwysocki the "types" export in notification hub package.json seem incorrect?

---

They resolve just fine in TypeScript as is.  I got the advice from @DanielRosenwasser and it works.

---

@mpodwysocki does TS not need this to resolve to a d.ts then? I guess I'm confused how the resolution works, lmk if you can point me at some docs

---

I am surprised too. The official doc example only shows using d.ts files 
https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing

---

Ugh. Maybe I got myself confused by thinking that the file extension should match whatever goes in an `import` or `require` field.

If I wrote those in the docs originally, they are likely the right way to do it; I must have had a lapse in judgment, but CC @weswigham @andrewbranch @RyanCavanaugh who might have other thoughts since they've been much more active in the area.

---

(I'm sorry for the churn @mpodwysocki)

---

Rule of thumb: Every code file should have a unique declaration file paired with it.

---

How's that weigh for this situation though - should the `types` field point to the `.d.ts` file or the `.js` file?

---

There should be two .d.ts files. One for the esm module, one for the cjs module. The esm module can just re-export stuff from the cjs one if it helps.

---

And, correspondingly, there should be two `types` fields. One for `require`, one for `import`. (Nested underneath)

---

The idea is basically that when you have both `require` and `import` conditions that point at separate files, you essentially have two separate, unique packages occupying the same package name (but with differing formats), and those two separate packages need declaration files with matching formats.

---

@weswigham It sounds like the documentation is incorrect then? Can you show what it would look like to nest `types` under each key, since I thought `require` and `import` mapped to simple strings rather than objects?

---

You can actually nest conditions arbitrarily deep, and use arrays of fallbacks, rather than just simple strings! They're quite complicated! Anyways, it would look something like
```json
"exports": {
  ".": {
    "import": {
        "types": "./types/index.d.ts",
        "default": "./dist/index.js"
    },
    "require": {
        "types": "./types/index.d.cts",
        "default": "./dist/index.cjs"
    }
  }
}
```

---

TIL! 😮 

Thanks for taking the time to explain, hopefully once we find the right config we can tool it such that humans don't have to think too hard about this stuff. 👍

---

> There should be two .d.ts files. One for the esm module, one for the cjs module. The esm module can (and probably should) just re-export stuff from the cjs one if it helps.

@weswigham thanks for the detail information! For our packages we used a rollup d.ts for the default `types` field. From my testing it worked when both `require` and `import` use the same d.ts file.  Is it necessary to have two separate files?

---

okay, from the linked discussion https://github.com/microsoft/TypeScript/issues/50466 we would need two separate files so that the correct mode is applied.

---

I updated this PR to have two copies of the same shim type file, `d.ts` for import, and `d.cts` for require. I tested the package on two projects, an ESM one with "type": "module", and a CJS one without "type": "module".  Result of VS Code IntelliSense/Go to Definition work the same as before so I assume this should be good.